src/librustc_typeck/check/_match.rs

```diff
@@ -212,7 +216,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             let mt = ty::TypeAndMut { ty: inner_ty, mutbl: mutbl };
                             let region = self.next_region_var(infer::PatternRegion(pat.span));
                             let rptr_ty = tcx.mk_ref(region, mt);
-                            self.demand_eqtype(pat.span, expected, rptr_ty);
+                            let err = self.demand_eqtype_diag(pat.span, expected, rptr_ty);
+                            if let Some(mut err) = err {
+                                if is_arg {
+                                    if let Ok(snippet) = self.sess().codemap()
+                                        .span_to_snippet(pat.span)
+                                    {
+                                        err.help(&format!("did you mean `{}: &{}`?",
```

Comment:
I tried adding these cases to the test and the results where:

```
fn ugh(&[bar]: &u32) {
}

error: slice pattern syntax is experimental (see issue #23121)
  --> ../../src/test/ui/mismatched_types/issue-38371.rs:26:9
   |
26 | fn ugh(&[bar]: &u32) {
   |         ^^^^^
```
```
fn agh(&&bar: &u32) {
}

error[E0308]: mismatched types
  --> ../../src/test/ui/mismatched_types/issue-38371.rs:29:9
   |
29 | fn agh(&&bar: &u32) {
   |         ^^^^ expected u32, found reference
   |
   = note: expected type `u32`
   = note:    found type `&_`
```

without changes to the code as is.